### DISCUSSION TITLE
Can GreptimeDB function as a relational database?

### DISCUSSION THREAD

**Discussion by skyformat99** [GreptimeTeam/greptimedb#4282] [Ideas]
都支持mysql访问了，好像增删改查也和mysql一样一样的啊
那么
能不能自己当作关系型数据库存储数据呢？
这样的话，时序数据，关系型数据都可以支持了

--- 
It already supports MySQL access, and it seems that adding, deleting, modifying, and querying are just like MySQL.
So, can it be used as a relational database to store data?
In this way, both time-series data and relational data can be supported

**Comment by killme2008** [comment]
感谢您的关注和问题。以下是对问题的回答：

看你是否需要一些关系数据库的关键特性，比如：
1. 外键和约束，GreptimeDB 仅支持主键约束（也就是时序的 tag）
2. 事务， GreptimeDB 当前不支持事务
3. 存储过程， GreptimeDB 当前也不支持存储过程调用
4. 对 SQL 标准的全面支持
5. 更完整和丰富的数据类型支持

以及 GreptimeDB 本身作为面向广义时序的一些约束，比如一定需要一个 time index 等。

如果以上这些问题对你来说都是可以接受的，确实是可以将 GreptimeDB 在一些场景下替代关系数据库使用，比如存储一些读多写少的元信息之类， time index 也可以通过默认赋值来绕过。

上面这些问题，随着 GreptimeDB 的发展，也会逐步完善，并且提供一些替代方案，比如存储过程，我们可能通过支持 python script 来解决（当前的支持不完善，将会重写）。但是 GreptimeDB 的定位还是一个希望能高效支持存储、分析、查询广义时序数据（事件、指标、日志、链路）的数据库，面向实时和海量数据处理来优化，不会也不应该变成一个关系数据库。


---------------------------------------------------------------------------

Check if you need some of the key features of relational databases, such as:

* Foreign keys and constraints: GreptimeDB only supports primary key constraints (which are the tags in time series).
* Transactions: GreptimeDB currently does not support transactions.
* Stored procedures: GreptimeDB also does not currently support stored procedure calls.
* Comprehensive support for SQL standards.
* More complete and rich data type support.

And some constraints of GreptimeDB itself as a time-series-oriented database, such as the requirement for a time index, etc.

If the above issues are acceptable to you, it is indeed possible to use GreptimeDB as a substitute for relational databases in some scenarios, such as storing some read-heavy and write-light metadata. The time index can also be bypassed by assigning a default value.

These issues will gradually be improved as GreptimeDB evolves, and some alternative solutions will be provided. For example, we may solve the issue of stored procedures by supporting Python scripts (current support is incomplete and will be rewritten). However, GreptimeDB is still intended to be a database that aims to efficiently support the storage, analysis, and querying of wide time-series data (events, metrics, logs, traces), optimized for real-time and massive data processing, and will not and should not become a relational database.
